internal_worker_base.go

```diff
@@ -121,18 +133,128 @@ func (bw *baseWorker) Start() {
 		return
 	}
 
-	// Launch the routines to do work
-	for i := 0; i < bw.options.routineCount; i++ {
+	for i := 0; i < bw.options.pollerCount; i++ {
 		bw.shutdownWG.Add(1)
-		go bw.execute(i)
+		go bw.runPoller()
 	}
+	bw.shutdownWG.Add(1)
+	go bw.runTaskDispatcher()
 
 	bw.isWorkerStarted = true
 	traceLog(func() {
-		bw.logger.Info("Started Worker", zap.Int("RoutineCount", bw.options.routineCount))
+		bw.logger.Info("Started Worker",
+			zap.Int("PollerCount", bw.options.pollerCount),
+			zap.Int("MaxConcurrentTask", bw.options.maxConcurrentTask),
+			zap.Int("MaxTaskRps", bw.options.maxTaskRps),
+		)
 	})
 }
 
+func (bw *baseWorker) isShutdown() bool {
+	select {
+	case <-bw.shutdownCh:
+		return true
+	default:
+		return false
+	}
+}
+
+func (bw *baseWorker) runPoller() {
+	defer bw.shutdownWG.Done()
+
+	bw.pollerRequestCh <- struct{}{} // initial poll request to kick off the poller loop
```

Comment:
I thought about that. The problem is that it won't work nicely with rate limiter. In the case of low RPS with high maxConcurrentTask, it will end up polling down a bunch of tasks that will only be blocked by rate limiter. The dispatcher does not actually process the task, it just receive from the channel and start a new routine to process it. With current setup, as soon as poller sends the task to the channel, dispatcher will receive it and it will send signal to pollerRequestCh immediately after which will cause the poller to continue polling so there is no blocking at all.